consensus.go

```diff
@@ -0,0 +1,181 @@
+package dcrlibwallet
+
+import (
+	"fmt"
+	"sort"
+
+	w "decred.org/dcrwallet/v2/wallet"
+	"github.com/decred/dcrd/chaincfg/chainhash"
+	"github.com/decred/dcrd/chaincfg/v3"
+)
+
+// SetVoteChoice sets a voting choice for the specified agenda. If a ticket
+// hash is provided, the voting choice is also updated with the VSP controlling
+// the ticket. If a ticket hash isn't provided, the vote choice is set for all
+// tickets controlled by the specified VSP.
+func (wallet *Wallet) SetVoteChoice(vspHost string, vspPubKey []byte, agendaID, choiceID, hash string, passphrase []byte) error {
+	wal := wallet.Internal()
+
+	if vspHost == "" && hash == "" {
+		return fmt.Errorf("request requires either a vsp host or a ticket hash")
+	}
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	// The wallet will need to be unlocked to sign the API
+	// request(s) for setting this vote choice with the VSP.
+	err := wallet.UnlockWallet(passphrase)
+	if err != nil {
+		return translateError(err)
+	}
+	defer wallet.LockWallet()
+
+	ctx := wallet.shutdownContext()
+
+	// get choices
+	choices, _, err := wallet.Internal().AgendaChoices(ctx, ticketHash) // returns saved prefs for current agendas
+	if err != nil {
+		return err
+	}
+
+	currentChoice := w.AgendaChoice{
+		AgendaID: agendaID,
+		ChoiceID: "abstain", // default to abstain as current choice if not found in wallet
+	}
+
+	for i := range choices {
+		if choices[i].AgendaID == agendaID {
+			currentChoice.ChoiceID = choices[i].ChoiceID
+			break
+		}
+	}
+
+	newChoice := w.AgendaChoice{
+		AgendaID: agendaID,
+		ChoiceID: choiceID,
+	}
+
+	_, err = wal.SetAgendaChoices(ctx, ticketHash, newChoice)
+	if err != nil {
+		return err
+	}
+
+	// If ticket hash is provided, then set vote choice for the selected ticket
+	// with the associated vsp.
+	if ticketHash != nil {
+		vspTicketInfo, err := wal.VSPTicketInfo(ctx, ticketHash)
+		if err != nil {
+			return err
+		}
+		// Update the vote choice for the ticket with the vsp
+		vspClient, err := wallet.VSPClient(vspTicketInfo.Host, vspTicketInfo.PubKey)
+		if err != nil {
+			return err
+		}
+		err = vspClient.SetVoteChoice(ctx, ticketHash, newChoice)
+		if err != nil {
+			// Updating the agenda voting preference with the vsp failed, therefore
+			// revert the locally saved voting preference for the agenda.
+			_, err = wal.SetAgendaChoices(ctx, ticketHash, currentChoice)
```

Comment:
This will overwrite the initial error from `vspClient.SetVoteChoice`; we don't want that. Either just do `wallet.Internal().SetAgendaChoices(ctx, ticketHash, currentChoice)` without assigning the error OR assign the error to a different variable and log the error if non-nil.